tests/helpers/test_template.py

```diff
@@ -1162,34 +1161,86 @@ def test_as_datetime(hass: HomeAssistant, input) -> None:
     )
 
 
-def test_as_datetime_from_timestamp(hass: HomeAssistant) -> None:
-    """Test converting a UNIX timestamp to a date object."""
-    tests = [
+@pytest.mark.parametrize(
+    ("input", "output"),
+    (
         (1469119144, "2016-07-21 16:39:04+00:00"),
         (1469119144.0, "2016-07-21 16:39:04+00:00"),
         (-1, "1969-12-31 23:59:59+00:00"),
-    ]
-    for input, output in tests:
-        # expected = dt_util.parse_datetime(input)
-        if output is not None:
-            output = str(output)
+    ),
+)
+def test_as_datetime_from_timestamp(
+    hass: HomeAssistant,
+    input: int | float,
+    output: str,
+) -> None:
+    """Test converting a UNIX timestamp to a date object."""
+    assert (
+        template.Template(f"{{{{ as_datetime({input}) }}}}", hass).async_render()
+        == output
+    )
+    assert (
+        template.Template(f"{{{{ {input} | as_datetime }}}}", hass).async_render()
+        == output
+    )
+    assert (
+        template.Template(f"{{{{ as_datetime('{input}') }}}}", hass).async_render()
+        == output
+    )
+    assert (
+        template.Template(f"{{{{ '{input}' | as_datetime }}}}", hass).async_render()
+        == output
+    )
 
-        assert (
-            template.Template(f"{{{{ as_datetime({input}) }}}}", hass).async_render()
-            == output
-        )
-        assert (
-            template.Template(f"{{{{ {input} | as_datetime }}}}", hass).async_render()
-            == output
-        )
-        assert (
-            template.Template(f"{{{{ as_datetime('{input}') }}}}", hass).async_render()
-            == output
-        )
-        assert (
-            template.Template(f"{{{{ '{input}' | as_datetime }}}}", hass).async_render()
-            == output
-        )
+
+@pytest.mark.parametrize(
+    "input",
+    (
+        "today_at('16:00')",
+        "today_at('16:00').date()",
+    ),
+)
+def test_as_datetime_from_datetime(hass: HomeAssistant, input) -> None:
+    """Test converting a timestamp string to a date object."""
+    today = dt_util.start_of_local_day()
```

Comment:
I'd create a constant at the module level for a datetime instead. We aren't interested specifically in the `start_of_local_day` datetime or using the `today_at` template function. We want to test the `as_datetime` template function. Then it will be easier to set a parameter for the expected result.

Reply:
I was struggling with this one.
I want to test with a datetime object as input (so not a datetime string). But if create a datetime object, and use it as input, it will use the python representation, so something like `datetime.datetime(2024, 1, 1, 0, 0, 0)` and then the tests will throw an error that it gets an integer where a comma is expected.

That's why I create a datetime object in the template itself.
I did update the test. I fixed the date to 1st of January 2024 so the output can be expected.
I could also use `as_datetime` instead of `today_at` but then I would be using the function I want to test in the test itself.

`today_at` and `as_datetime` are the only options I could think of which give predictable output. `now()` will change while the test is ongoing.

example with a datetime.datetime as input:
```log
self = <homeassistant.helpers.template.TemplateEnvironment object at 0x7f06bfd89450>, source = '{{ as_datetime(2024-01-05 13:22:15.862657) }}'

    def handle_exception(self, source: t.Optional[str] = None) -> "te.NoReturn":
        """Exception handling helper.  This is used internally to either raise
        rewritten exceptions or return a rendered traceback for the template.
        """
        from .debug import rewrite_traceback_stack
    
>       raise rewrite_traceback_stack(source=source)
E       jinja2.exceptions.TemplateSyntaxError: expected token ',', got 'integer'

/usr/local/lib/python3.11/site-packages/jinja2/environment.py:936: TemplateSyntaxError

The above exception was the direct cause of the following exception:

hass = <HomeAssistant RUNNING>, input = datetime.datetime(2024, 1, 5, 13, 22, 15, 862657), output = '2024-01-01 16:00:00-08:00'

    @pytest.mark.parametrize(
        ("input", "output"),
        [
            (
                "today_at('16:00').replace(year=2024, month=1, day=1)",
                "2024-01-01 16:00:00-08:00",
            ),
            (
                "today_at('16:00').replace(year=2024, month=1, day=1).date()",
                "2024-01-01 00:00:00",
            ),
        ],
    )
    def test_as_datetime_from_datetime(
        hass: HomeAssistant, input: str, output: str
    ) -> None:
        """Test converting a timestamp string to a date object."""
    
        input = datetime.now()
>       assert (
            template.Template(f"{{{{ as_datetime({input}) }}}}", hass).async_render()
            == output
        )

tests/helpers/test_template.py:1215: 
```

Reply:
Changed it to use `as_datetime` for the input